[1단계 - 로또 구현] 마크(김병화) 미션 제출합니다.

binghe819 opened the pull request:
### 안녕하세요 hue :)

<br>

이번 미션에서는 일급  컬렉션을 만드는 것과 TDD에 집중해서 코드를 작성했어요!!

피드백을 받는 건 언제든 기대되는군요!ㅎㅎ 리뷰를 통해 성장을 희망해요! 따끔한 피드백들 부탁해요 hue :)

<br>

## 궁금한 점

<br>

### 컨트롤러의 역할
저는 계속해서 MVC패턴을 구현하면서 입력 관련해서 View에서 그저 `String` 만을 입력받아 `Controller` 에 넘겨주고, `Controller`에서 예외처리와 동시에 숫자나 특정 객체로 변환을 했는데요!

그전까지는 입력값을 받는 것이 많지 않아서 `Controller`의 코드가 비교적 짧고 책임도 명확했는데요! 

로또 미션을 진행하면서는 입력값을 받아 예외처리와 특정 객체로 변환할 일이 많아서 `Controller`의 코드가 길어지고, 추후엔 책임도 점점 많아질 것이란 생각이 들어서 질문을 남깁니다!

* `Controller`의 역할은 객체를 원활하게 협력하도록 위임하는 책임이 가장 크다고 생각하는데요! **입력 값이 많을 수록 위임도 하지만 입력값 예외처리를 하는 책임이 더 크다고 느껴지는데요! 이게 `Controller` 의 역할이 맞는지 궁금해요!**
* 지금까지는 `Controller`가 하나 뿐인 프로그램을 만들었는데, 프로그램의 범위가 커지면 `Controller` 의 책임을 분리시키긴 위해선 `Controller` 가 여러개로 나눠질 수도 있다고 느껴져요! **`Controller` 를 여러개로 분리시키는 기준이 있는지 궁금해요!**

<br>

### TDD
이번 미션을 진행하며 아직 잘 모르지만 TDD 방식으로 코드를 작성해나갔는데요! 

TDD로 코드를 작성하며 궁금한 점이 생겨 질문 남깁니다!

* 이번 미션을 진행하며 작은 범위의 도메인인 `LottoNumber` (하나의 정수를 포장하고 있는 객체) 부터 TDD로 진행했는데요! 점점 큰 범위의 도메인을 작성할 수록 객체가 어떤 책임을 가져야하는지 모호해지더라구요! 그래서 저는 중간에 MVC구조를 만들고 `Controller` 로직을 작성하고나서 나머지 도메인을 작성했는데요! **hue는 어떤 순서로 TDD를 진행하는지 궁금해요!!**
* **TDD로 코드를 작성하기 전에 도메인 모델(예를 들어 UML 형식으로 작성)을 어느정도까지 설계하고 코드를 작성하는지 궁금합니다!** 
  * 정말 간단하게만 설계하고 TDD로 코드를 작성해나가니 테스트 가능하고 유연한 구조가 나오는 반면, 어느 순간부터는 막막해지더라구요ㅠ

<br>

### 예외처리 전략
이전까지는 모든 원시값과 문자열을 포장하지 않았고, 특정 시점에서만 예외처리를 했어서 예외를 던지는 곳이 한정적이었는데요!

* 모든 원시값과 문자열을 포장하면, 객체 생성시 예외처리 로직도 포함되기때문에 만약 악의를 가진(?) 클라이언트가 예외를 발생시킨다면 어떻게 예외를 처리할지 궁금하더군요!
  * 물론 대부분의 예외는 `Controller`에서 잡아내지만, 포장 객체에서 Checked Exception를 던지게 만들 수도 없기 때문에 `Controller`가 포장된 객체를 생성하는 모든 코드를 커버하진 못한다고 생각이 들어서 질문드립니다!

<br>

더 궁금한 점들이 생기면 DM으로 연락드릴게요(괴롭혀드릴게요!)!ㅎㅎ

binghe819 committed:
refactor: 사용하지 않는 파일 삭제

binghe819 committed:
docs: 기능 명세 작성

binghe819 committed:
feat: 로또 구입 금액 입력 기능

binghe819 committed:
feat: 로또 번호 생성 기능

binghe819 committed:
refactor: 로또 번호 생성 기능 리팩토링

binghe819 committed:
feat: 당첨 번호 입력 기능

binghe819 committed:
feat: MVC 패턴 구조 설계 및 로또 구입 금액 입력 기능 구현

binghe819 committed:
refactor: 패키지 정리 및 포맷팅

binghe819 committed:
test: 랜덤 로또 번호 생성 테스트 코드 추가

binghe819 committed:
refactor: 로또 구입 금액 입력 리팩토링

binghe819 committed:
docs: 기능 명세 추가 및 구현한 기능 체크

binghe819 committed:
feat: 로또 구입 금액에 맞는 로또 생성 기능

binghe819 committed:
refactor: 패키지 정리

binghe819 committed:
feat: 당첨 번호 입력 및 보너스 볼 입력 기능 구현

binghe819 committed:
feat: 당첨 확인 및 수익률 확인 기능 구현

binghe819 committed:
refactor: 포맷팅

binghe819 committed:
refactor: 당첨 결과 정렬해서 출력하도록 리팩토링

binghe819 committed:
refactor: 하드코딩된 값 상수처리

binghe819 commented:
# [JDK] Stream API - 1
## 내용
- Stream API에 대해서 공부하고 활용하기 위해 노력함
- Stream API를 최대한 활용하여 데이터를 가공하고 가독성을 높일 수 있었음
## 링크
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-b759f2ba9568ddd596e929a75a96ea38949f2c36cbd62caf86e4a467f7f0ee09
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-262a4a843b87d56ccd455cce2cc72db8d72a5edea3eea3f1a191dd3695861fba
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-6171e15ae725140ecd4dd6b5660c742a39cde4d80d3698bfb9b562ec290b6903
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-a188d52585281448755e6fbd05c60738529a3887ad3e771c5271be7bc6de4f56
- [Stream 정리내용](https://github.com/binghe819/TIL/blob/master/JAVA/%EB%AA%A8%EB%8D%98%20%EC%9E%90%EB%B0%94/Stream.md)

<br>

# [OOP] 원시값 포장 - 4
## 내용
- `LottoNumber` 객체를 도출하여 `모든 원시값과 문자열을 포장함` 규칙을 지킴
- `LottoNumber`와 관련된 비즈니스 로직을 `Lotto`에서 분리할 수 있었음
- 이외에도 원시값 포장을 최대한 활용하여 자율적인 객체를 만들 수 있었음
## 링크
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-7b3cad6cd6baa0c33143a96f842df63abf241a7b8b9a29f7498743ce89d0e2ac
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-262a4a843b87d56ccd455cce2cc72db8d72a5edea3eea3f1a191dd3695861fba
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-fa6c919752f9904a6f9f6d4d44c4dfcf571854cc6bc4c43c7d302bc51b12c8a0

<br>

# [JDK] Enum - 3
## 내용
- `LottoRank` (당첨 순위)관련해서 `if` 문을 사용하지 않고 Enum을 이용해 당첨 개수를 보내면 개수에 맞는 당첨 순위 Enum 객체를 반환하도록 함
- Enum의 기초를 정리함
## 링크
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-717a9978b8613f482fc52c001ee5b5c3d66186d6f16c35aca896b47804808725
- [Enum 기초정리](https://github.com/binghe819/TIL/blob/master/JAVA/%EA%B8%B0%ED%83%80/%EC%97%B4%EA%B1%B0%ED%98%95(enum).md)

<br>

# [TDD] 테스트코드 - 5
## 내용
- 실패하는 테스트 코드를 먼저 작성하고, 통과시키도록 프로덕트 코드를 만든 후, 리팩토링을 진행함.
- 모든 객체에 대한 단위테스트를 하기 위해 노력함
## 링크
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-081b1b18f654fafed8d6b46995d5e2bbdcfdcd28bb2fe4f2e5d76efd0912a7f5
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-79daa40c2112c060139561ff09bdaed90607a2543390c00fefdbd00de19a4cc2
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-6e5b7800a367755524e824c7822d0ca206ec9153b65f564b89e8f14fcc48dc9d

<br>
 
# [OOP] MVC - 3
## 내용
- MVC 구조로 각자의 역할을 가지고 서로 협력하도록 노력함.
- `View`의 요구사항에 따라 출력하도록, `Controller`에서는 데이터만 `View`에 넘기고, `View`의 요구사항에 따라 출력하도록 함. (MVC 서로 의존성을 최대한 줄임)
## 링크
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-940d1c44120ea83f932b0ee11c19b877bf63456804f1092acf8a622ed268e26c

<br>

# [JDK] Optional - 1
## 내용
- null 처리를 Optional을 이용해서 처리함.
## 링크
- https://github.com/woowacourse/java-lotto/pull/263/files#diff-717a9978b8613f482fc52c001ee5b5c3d66186d6f16c35aca896b47804808725
- [Optional 정리 글](https://github.com/binghe819/TIL/blob/master/JAVA/%EB%AA%A8%EB%8D%98%20%EC%9E%90%EB%B0%94/Optional.md)

Hue9010 commented on src/main/java/lotto/domain/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,47 @@
+        this.value = value;
+    }
+
+    public static LottoNumber valueOf(int value) {
+        validateLottoNumber(value);
+        return new LottoNumber(value);
+    }
```
> 예외처리 전략
> 이전까지는 모든 원시값과 문자열을 포장하지 않았고, 특정 시점에서만 예외처리를 했어서 예외를 던지는 곳이 한정적이었는데요!
> 
> - 모든 원시값과 문자열을 포장하면, 객체 생성시 예외처리 로직도 포함되기때문에 만약 악의를 가진(?) 클라이언트가 예외를 발생시킨다면 어떻게 예외를 처리할지 궁금하더군요!
>    - 물론 대부분의 예외는 Controller에서 잡아내지만, 포장 객체에서 Checked Exception를 던지게 만들 수도 없기 때문에 Controller가 포장된 객체를 생성하는 모든 코드를 커버하진 못한다고 생각이 들어서 질문드립니다!

클라이언트라고 표현해주셔서 http로 통신하는 client-server에서의 예시를 알려드리자면 Controller에서 잡았던 LttoNumber 생성자에서 잡았던 bad request(status code 400)를 응답해 줍니다. 그럼 이를 밭은 클라이언트는 경우에 따라 사용자한테 잘못된 값을 입력했다는 팝업을 보여 줄 수 있을 것 같네요.
어디서 예외가 발생했는지가 중요한게 아니라 어떤 예외가 발생했는지가 중요하겠죠? 😉 

확실히 답변이 됐는지 모르겠네요. 😢  아니라면 상정하신 케이스를 알려주시면 좋을 것 같아요.

Hue9010 commented on src/test/java/lotto/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,81 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class WinningLottoTest {
```
> TDD
> 이번 미션을 진행하며 아직 잘 모르지만 TDD 방식으로 코드를 작성해나갔는데요!
> 
> TDD로 코드를 작성하며 궁금한 점이 생겨 질문 남깁니다!
> 
> - 이번 미션을 진행하며 작은 범위의 도메인인 LottoNumber (하나의 정수를 포장하고 있는 객체) 부터 TDD로 진행했는데요! 점점 큰 범위의 도메인을 작성할 수록 객체가 어떤 책임을 가져야하는지 모호해지더라구요! 그래서 저는 중간에 MVC구조를 만들고 Controller 로직을 작성하고나서 나머지 도메인을 작성했는데요! hue는 어떤 순서로 TDD를 진행하는지 궁금해요!!
> - TDD로 코드를 작성하기 전에 도메인 모델(예를 들어 UML 형식으로 작성)을 어느정도까지 설계하고 코드를 작성하는지 궁금합니다!
정말 간단하게만 설계하고 TDD로 코드를 작성해나가니 테스트 가능하고 유연한 구조가 나오는 반면, 어느 순간부터는 막막해지더라구요ㅠ

요구사항을 정리하고 그걸 기반으로 객체를 뽑아내고 해당 객체에 필요한 기능을 리스트업 합니다.(유효성 검증 포함) 그리고 그걸 기반으로 작업을 시작합니다. 당연히 스펙화 시킨 내용이 구현 도중에 변경되기도 합니다. 😏

그러나 위는 기존 프로젝트에 기능을 추가할 때의 얘기니까 지금처럼 처음부터 만들 때는 조금 얘기가 달라집니다.ㅎㅎ
지금처럼 망망대해에 추가해야할 로직이 많다면 대략적인 요구사항을 정리 해둔 다음 가장 단순한 형태로 실행 시키기 위해 필요한 구현 순서대로 요구사항을 정리해두고 이에 맞춰 마크처럼 한 곳(예로 Controller)에서 작업 해가면서 그때 그때 필요한 객체를 추가해 나갈 것 같습니다.

사내 시니어 분께 전해듣기로는 크게 두가지 타입이 있다고 하더군요. 처음부터 설계를 꼼꼼히 해 두고 그에 맞춰서 작업하는 스타일, 우선 만들고 잘못 된 곳을 빠르게 고치는 스타일
어느게 좋다 나쁘다가 아니라 사람마다 스타일이 다른 것 뿐이기 때문에 자신의 스타일을 찾아가는 것도 좋다고 생각해요. 😉

Hue9010 replied:
직업인으로서의 개발자는 요구사항을 코드로 구현해주는 사람이라고 생각합니다. 다양한 직업 가치가 있겠지만 돈을 받는 입장에서 결국 개발자도 직업인의 하나라고 생각해요. 그러면 회사가 원하는 핵심 가치는 회사의 요구사항을 구현해주는 사람이 아닐까 싶어요.
이런 얘기를 왜 하냐하면! 꼭 회사가 아니더라도 혼자 관리하고 재미로 만드는 프로젝트라고 하더라도 결국엔 자신의 요구사항을 코드화 시키는 것입니다. 즉, 그만큼 요구사항이라는게 중요합니다.

요구사항을 정리하고, 구현이 되어야 하는 순서대로 정렬 해둔다음 작업해보면 어떨까요?
이건 어디까지나 저의 방법이고 마크만의 방법을 찾아가는게 가장 베스트입니다. 😉 
요구사항의 정리는 꼭 README일 필요는 없어요. 개인적으로는 노트북 옆 노트에 수기로 작성하고 필요할 때마다 수정하는걸 선호합니다.(그러긴해도 과정에서 권장하니까 README는 챙겨보시길 추천드립니다. 😋 )

Hue9010 commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,33 @@
+
+    public LottoNumber getBonus() {
+        return bonus;
+    }
```
사용하지 않게된 메서드는 그때 그때 정리해 주세요!
이번 작업을 하면서도 사용하지 않는 메서드가 생겼는데 수년 째 운영 중인 프로젝트에서 사용하지 않는 코드들을 제거하지 않고 있다면 어떨까요? 😱

Hue9010 commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,33 @@
+        boolean isMatchBonus = userLotto.getNumbers().contains(bonus);
+        long numOfMatch = userLotto.getNumbers()
+            .stream()
+            .filter(lottoNumber -> winningLotto.getNumbers().contains(lottoNumber))
```
getNumbers를 가져와서 contains로 확인하기보다는 WinningLotto에서 contain을 제공해 주면 어떨까요?

Hue9010 commented on src/main/java/lotto/model/LottoRank.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public static LottoRank valueOf(long numOfMatch, boolean isMatchBonus) {
+        if (numOfMatch == 5 && isMatchBonus) {
```
5가 매직 넘버로 보이네요. 🙄 
그렇다고 SECOND.correct, THIRD.correct 한개를 사용하기에는 둘 다 포함된 의미라 별도의 상수롤 빼주는게 가장 깔끔해 보여요.

Hue9010 commented on src/main/java/lotto/model/LottoRank.java:
```diff
@@ -0,0 +1,43 @@
+        return Arrays.stream(LottoRank.values())
+            .filter(lottoRank -> lottoRank.getCorrect() == numOfMatch)
+            .findFirst()
+            .orElse(NO_MATCH);
```
orEles 👍 
Optional까지 잘 사용해 주셨네요. 💯

Hue9010 commented on src/main/java/lotto/model/LottoResult.java:
```diff
@@ -0,0 +1,47 @@
+        if (result.containsKey(lottoRank)) {
+            int num = result.get(lottoRank);
+            result.put(lottoRank, num + 1);
+        }
```
[자바 HashMap을 효과적으로 사용하는 방법](http://tech.javacafe.io/2018/12/03/HashMap/)
computeIfPresent 😉

Hue9010 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
```
사용하지 않네요. 😉 
```
Import 정리하기 (Optimize Imports)
MacOS: Ctrl + Opt + o
Win/Linux: Ctrl + Alt + o
자동 설정: [Find Action] -> Optimize imports on 입력 -> “Auto import: …”
```
[[IntelliJ] intellij 유용한 단축키 정리](https://gmlwjd9405.github.io/2019/05/21/intellij-shortkey.html)

Hue9010 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,89 @@
+
+public class LottoController {
+
+    private final InputView inputView;
```
Controller는 일반적으로 상태를 가지지 않아요. OutputView와 별개로 InputView만 Controller의 상태로 만든 이유가 있나요? 🤔

binghe819 replied:
프리코스 미션에서 main함수에 Scanner 인스턴스를 생성하여 구현하라는 요구사항에 익숙해져서 자연스럽게 이렇게 만든 것 같아요ㅠ

지금보니 이렇게 코드를 작성할 이유가 없네요ㅎㅎ 좋은 지적 감사합니다 :)

Hue9010 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Scanner;
+import java.util.stream.Collectors;
+
+public class InputView {
```
Input이기 때문에 정말 입력에 대한 처리만 넣은걸로 보입니다.
MVC의 V를 웹페이지로 생각해볼까요?
회원가입 페이지는 입력화면이면서 회원 가입에 필요한 내용을 사용자한테 출력해주겠죠?
이에 맞춰 InputView라고 해서 정말 입력만 받지 말고 입력에 필요한 출력은 InputView가 담당하게 해줄 수 있을 것 같아요.
출력을 배제하고 만들면 View관련 코드가 굉장히 복잡해지는 상황이 생기죠ㅎㅎ

View를 웹페이지에 대입해서 생각하면 다른 것도 View에서 가능하답니다.
회원가입을 다시한번 생각해보면 잘못된 비밀번호를 입력했을 때 서버에 요청 없이도 페이지 내에서 유효성 검사를 하기도 합니다.
InputValidationUtils을 없앨 수 있겠군요. 😉 

나중 스프링 환경이라면 웹 페이지(백엔드 과정이기때문에 선택적으로), Controller, Serivce, Domain에서 일부 중복 된 유효성 검사를 할 수도 있습니다. 그리고 각 단계에서의 유효성 검사는 동일한 검사라도 서로 의미가 다릅니다.
웹페이지는 사용자가 유효한 값을 입력했는지, Controller는 클라이언트가 유효한 값을 전달 했는디, 객체는 생성 시 유효한 값이 왔는지 확인하는게 될 것 같네요. 결국 같은 검사라해도 각자의 위치에서는 다른 의미를 가집니다.

그러나 지금 단계에서 그렇게까지 꼼꼼한 유효성 검사를 하는검 추천드리지 않습니다. 이유는 코드가 지나치게 복잡해지고 집중해야 할 다른 것보다 지나치게 유효성 검사에 많은 시간을 할애하게 됩니다.
그리고 스프링 환경에서는 특정 애노테이션을 붙이느 것만으로도 간간한 유효성 검사가 가능하여 손쉽게 검사 할 수 있는데 지금은 직접 작성 해줘야 합니다. 😢 

결론적으로는 객체 중심의 유효성 검사에 집중하는 걸 추천드립니다. 객체에서 최종적으로 검사를 하게 되면 잘못된 객체가 생성될 일이 원천적으로 차단 됩니다!

Hue9010 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,89 @@
+            availableLotto.add(Lotto.generatedBy(new RandomLottoGenerator()));
+        }
+        return new Lottos(availableLotto);
+    }
```
금액 기준으로 구매할 로또 갯수를 구하고 그만큼 로또를 생성하는 로직이네요.
이를 담당할 매니저, 스토어, 팩토리 등등을 만들어 볼 수 있을 것 같네요. (혹은 Lottos가 정적 팩터리 메서드 제공)

Hue9010 replied:
별도 모델을 만든다면 inputWinningLotto, inputWinningBonus도 같이 옮길 수 있을 걸 것네요

Hue9010 commented on src/main/java/lotto/domain/lottogenerator/RandomLottoGenerator.java:
```diff
@@ -0,0 +1,30 @@
+    public List<LottoNumber> generateLottoNumbers() {
+        List<LottoNumber> lottoNumbers = getShuffledLottoAllNumbers()
+            .stream()
+            .map(number -> LottoNumber.valueOf(number))
```
메서드 참조를 사용할 수 있을 것 같네요. 😉 
[Java Lambda (3) 메소드 참조](https://futurecreator.github.io/2018/08/02/java-lambda-method-references/)

Hue9010 commented on src/main/java/lotto/domain/lottogenerator/RandomLottoGenerator.java:
```diff
@@ -0,0 +1,30 @@
+        List<Integer> lottoAllNumbers = new ArrayList<>();
+        for (int i = LottoNumber.MIN + 1; i < LottoNumber.MAX; i++) {
+            lottoAllNumbers.add(i);
+        }
```
한번 만들어 놓고 사용 할 수 있을 것 같은데 List를 매번 만들어 줄 필요가 있을지 생각 해보죠. 🤔 

다시보니 List\<LottoNumber>를 만들어 둬도 되지 않을까요? 🤔 222

Hue9010 commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+
+    private static void validateLottoNumberNotDuplicate(List<LottoNumber> lottoNumbers) {
+        List<LottoNumber> distinctedLottoNumbers = lottoNumbers.stream()
+            .distinct()
```
distinct 👍

Hue9010 commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+public class Lotto {
+
+    public static final int PRICE = 1000;
+    public static final int LENGTH = 6;
```
단순 LENGTH라기보다는 로또 길이가 맞을 것 같아요.
public인 이유는 RandomLottoGeneratorTest 때문일까요? 
테스트를 위해 접근 제어자를 바꾸지 말죠. 😉

Hue9010 commented on src/main/java/lotto/exception/LessThanLottoPriceException.java:
```diff
@@ -0,0 +1,8 @@
+public class LessThanLottoPriceException extends IllegalArgumentException {
+
+    public LessThanLottoPriceException() {
+        super("[ERROR] " + "로또 가격이 1000원이므로, 구입 가격은 1000원부터 시작입니다.");
```
더하기 없이 하나의 문자열로 만들 수 있을 것 같네요. 🙄

Hue9010 commented on src/main/java/lotto/exception/LessThanLottoPriceException.java:
```diff
@@ -0,0 +1,8 @@
+package lotto.exception;
+
+public class LessThanLottoPriceException extends IllegalArgumentException {
```
IllegalArgumentException 👍

Hue9010 commented on src/main/java/lotto/model/Money.java:
```diff
@@ -0,0 +1,56 @@
+            return true;
+        }
+        return false;
+    }
```
조건문을 만들 때 한두번씩 하기 쉬운 실수인데
```java
return this.price < price;
```
로 충분하지 않을까요? 😏

Hue9010 commented on src/main/java/lotto/model/Money.java:
```diff
@@ -0,0 +1,56 @@
+        return new Money(value);
+    }
+
+    private static void validateMoney(int value) {
+        if (value < 0) {
+            throw new IllegalArgumentException("금액은 마이너스가 될 수 없습니다.");
+        }
+    }
```
입력에서만 1000원에 대한 validation이 있고 Money자체는 1000원 이하에 대한 제약조건이 없게 만드신게 맞을까요?
그게 맞다면 Money라는 이름처럼 로또에 제한되지 않고 좀더 유연하게 사용 가능하겠네요. 👍

Hue9010 commented on src/main/java/lotto/model/Money.java:
```diff
@@ -0,0 +1,56 @@
+
+public class Money {
+
+    private static final Money ZERO = Money.priceOf(0);
```
사용하지 않는 건 그때 그때 지워주죠. 
전체적으로 한번 체크해보세요. 😉

Hue9010 commented on src/main/java/lotto/model/LottoResult.java:
```diff
@@ -0,0 +1,47 @@
+public class LottoResult {
+
+    private final Map<LottoRank, Integer> result;
+    private int totalPurchaseAmount = 0;
```
생성자가 있으니 생성자에서 초기화 해주죠. 😉

Hue9010 commented on src/main/java/lotto/domain/lottogenerator/LottoGenerator.java:
```diff
@@ -0,0 +1,10 @@
+import java.util.List;
+import lotto.domain.lotto.LottoNumber;
+
+@FunctionalInterface
```
FunctionalInterface 👍

Hue9010 commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+        List<LottoNumber> lottoNumbers = numbers.stream()
+            .map(number -> LottoNumber.valueOf(number))
+            .collect(Collectors.toList());
+        return Lotto.generatedBy(() -> lottoNumbers);
```
😮 
함수형 인터페이스 👍

Hue9010 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,89 @@
+        } catch (Exception e) {
+            OutputView.printExceptionMessage(e);
+            return inputWinningLotto();
+        }
```
> 저는 계속해서 MVC패턴을 구현하면서 입력 관련해서 View에서 그저 String 만을 입력받아 Controller 에 넘겨주고, Controller에서 예외처리와 동시에 숫자나 특정 객체로 변환을 했는데요!
> 
> 그전까지는 입력값을 받는 것이 많지 않아서 Controller의 코드가 비교적 짧고 책임도 명확했는데요!
> 
> 로또 미션을 진행하면서는 입력값을 받아 예외처리와 특정 객체로 변환할 일이 많아서 Controller의 코드가 길어지고, 추후엔 책임도 점점 많아질 것이란 생각이 들어서 질문을 남깁니다!
> 
> - Controller의 역할은 객체를 원활하게 협력하도록 위임하는 책임이 가장 크다고 생각하는데요! 입력 값이 많을 수록 위임도 하지만 입력값 예외처리를 하는 책임이 더 크다고 느껴지는데요! 이게 Controller 의 역할이 맞는지 궁금해요!
> - 지금까지는 Controller가 하나 뿐인 프로그램을 만들었는데, 프로그램의 범위가 커지면 Controller 의 책임을 분리시키긴 위해선 Controller 가 여러개로 나눠질 수도 있다고 느껴져요! Controller 를 여러개로 분리시키는 기준이 있는지 궁금해요!

그래서 가능하다면 복잡도를 높히기 위해 잘못 된 값을 입력했을 때 재입력 받는 로직 등등 없이 심플하게 만들어 보는걸 추천드리는 편입니다. 😉 (이 위치에 코맨트를 남기는 이유입니다ㅎㅎ)
하지만 이건 추천사항일 뿐 개선이 필요한 사항은 아닙니다.
예외에 대한건 다른 코맨트에도 남겼는데 Controller도 클라이언트가 정상적인 데이터를 넘겨줬는지 검증 할 수도 있습니다. 이는 나중에 스프링으로 넘어가면 애노테이션을 통해 손쉽게(?) 가능한 것도 있어서 지금 단계에선 Controller 단에서의 검증은 별도로 하지말고 Controller에서 할 법한 검증은 View에서 하고 객체 내의 검증에 집중하는 걸 추천드립니다.

지금의 애플리케이션은 DB가 없고 한번 실행 후 끝나는 등 몇가지 제약사항이 있습니다. 그러다 보니 실제 api server였다면 입력 갯수만큼 서로 다른 API를 요청 했을 텐데 제약 사항때문에 하나의 메서드로 모든걸 끝내버립니다. 그러다 보니 마크의 노력과 별개로 별수 없이 더 복잡해지는 것도 있습니다. 😢 
클라이언트와 상호작용을 하는 중심이 되는 도메인별로 Controller를 나눕니다. 
커뮤니티 사이트를 생각하면 BoardController, UserController가 될 것 같고 Comment는 댓글의 기능이 어느정도 중요한지에 따라 BoardController에서 같이 제공하거나 CommentConroller로 별도 분리 할 수 있을 것 같네요.

Hue9010 replied:
학습 로그에 작성하신 TIL들 봤는데 잘 정리해주셨네요. 👍👍👍
Java 9에서 Optinal에 추가된 기능도 확인해보시면 좋을 것 같아요. 😉 

[Java9에서 추가된 Optional 기능 소개](https://www.daleseo.com/java9-optional/)

Hue9010 reviewed (changes requested):
안녕하세요 마크! 이번 리뷰를 맡은 휴입니다! 😄 
잘 만들어 주신 덕분에 로또 플레이 잘 해봤습니다.😉 
질문에 대한 답변과 개선하면 좋을 부분 피드백 남겼습니다.(혹시 놓친 질문이 있다면 알려주세요! 🙏 )
궁금한건 언제든 코멘트나 DM으로 질문주세요!

binghe819 committed:
docs: 1단계 피드백 정리 및 준수한 프로그래밍 요구사항 체크

binghe819 committed:
refactor: Lotto가 특정 LottoNumber를 가지는지 확인하는 로직 리팩토링

binghe819 committed:
refactor: 로또 순위 객체와 로또 객체 상수화 및 네이밍 수정

binghe819 committed:
refactor: HashMap computeIfPresent 리팩토링

binghe819 committed:
refactor: 로또 번호 생성기의 숫자 리스트 캐싱 및 메서드 참조로 리팩토링

binghe819 committed:
refactor: 비효율적인 문자열 더하기 및 조건문 리팩토링

binghe819 committed:
refactor: Controller와 View의 책임 리팩토링

변경후의 책임
- Controller
View로부터 입력값 받아서 도메인 객체로 변환 + 전체적인 프로그램의 흐름 제어
예외처리 : 객체 중심 예외처리만 수행 (생성자를 통한 예외처리)

- View
입력값 예외처리하고 반환
예외처리 : 입력값 예외처리만 수행

binghe819 committed:
refactor: 금액 기준으로 구매할 로또 갯수를 구하고 로또를 생성하는 로직 분리

binghe819 committed:
feat: 보너스 볼이 당첨 번호에 포함되어있다면 예외처리하는 기능 구현

binghe819 committed:
refactor: 입력 번호와 당첨 번호의 당첨 개수 구하는 로직 리팩토링

- Lotto와 Lotto가 비교하여 몇 개의 같은 번호가 존재하는지 확인하도록 변경함 (자율적인 객체)

binghe819 committed:
refactor: 사용하지 않는 메서드 및 import 모두 제거

binghe819 committed:
refactor: 입력값에 대한 예외처리 리팩토링

binghe819 commented:
### 안녕하세요 휴! :)

<br>

휴의 정성이 담긴 피드백에 맞게 저도 최선을 다해 수정을 해보았어요 👍 

수정 내용은 [여기](https://github.com/binghe819/java-lotto/blob/step1/docs/README.md#1%EB%8B%A8%EA%B3%84-%ED%94%BC%EB%93%9C%EB%B0%B1-%EC%B2%B4%ED%81%AC%EB%A6%AC%EC%8A%A4%ED%8A%B8)에서 한눈에 확인가능하세요 :) 

여러가지 수정을 진행하였고, 그중에서도 자율적인 객체를 만들고, View와 Controller에 대해 깊게 생각해보고 수정해보았어요!

하지만, 제가 휴의 피드백 내용대로 수정을 했는지 정확히 몰라서 DM으로도 질문을 드렸지만, 휴가 직접 제가 수정한 코드를 보시고 고쳤으면 하는 부분에 대해서 콕콕 집어주시면 좋을 것 같아요!!ㅎㅎ

### ps. 급하게 진행하기보다는 여러가지로 씹고 뜯고 맛볼 예정이라 편한 시간에 편하게 피드백 남겨주세요ㅎㅎ

Hue9010 commented on docs/README.md:
```diff
@@ -0,0 +1,51 @@
+
+<br>
+
+## 1단계 피드백 체크리스트
```
꼼꼼하게 잘 챙기셨네요. 👍

Hue9010 commented on src/main/java/lotto/domain/lotto/Lotto.java:
```diff
@@ -0,0 +1,66 @@
+        return new ArrayList<>(numbers);
+    }
+
+    public boolean isContains(LottoNumber lottoNumber) {
+        return numbers.contains(lottoNumber);
+    }
```
간단한 기능이긴 해도 다음엔 TDD에 맞춰 테스트를 추가해보는 버릇을 들여보시죠. 
도리어 간단하니까 테스트를 먼저 만들어 보는것도 쉬울겁니다. 😉

Hue9010 commented on src/main/java/lotto/model/LottoResult.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public void addRank(LottoRank lottoRank) {
+        result.computeIfPresent(lottoRank, (LottoRank key, Integer value) -> ++value);
```
👍 
알아두면 편한 것들이 있답니다. 
몇몇개는 상황(지금처럼 카운팅을 하는 문제)에 따라 알고리즘 문제에서도 사용가능합니다. 😉

binghe819 replied:
1단계 피드백에서 알려주신 `HashMap 효과적으로 사용하는 방법`글을 보고 직접 정리하며 공부했는데 큰 도움이 되었습니다 👍

Hue9010 commented on src/test/java/lotto/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,94 @@
+        assertThat(lottoResult).isEqualTo(LottoRank.SECOND);
+    }
+
+    @DisplayName("보너스 번호는 당첨 번호는 없는 번호여야한다.")
```
테스트 추가 💯

Hue9010 commented on src/main/java/lotto/model/Money.java:
```diff
@@ -0,0 +1,48 @@
+        this.price = price;
+    }
+
+    public static Money priceOf(int value) {
+        validateMoney(value);
+        return new Money(value);
+    }
```
```
before
View
- 입력값 반환
- 예외처리 : X
Controller
- View로부터 입력값 받아서 예외처리하고 도메인 객체로 변환 + 전체적인 프로그램의 흐름 제어
- 예외처리 : 입력값 예외처리 + 객체 중심 예외처리 (생성자를 통한 예외처리)

after
View
- 입력값 예외처리하고 반환
- 예외처리 : 입력값 예외처리만 수행
Controller
- View로부터 입력값 받아서 도메인 객체로 변환 + 전체적인 프로그램의 흐름 제어
- 예외처리 : 객체 중심 예외처리만 수행 (생성자를 통한 예외처리)
```

Controller의 예외처리가 어떤걸 말씀하시는건가 싶었는데 이 부분을 말씀하신거군요.
이건 Controller의 예외처리라기보다는 각 모델의 자체적인 예외처리로 볼 수 있겠네요.

리뷰어마다 View에서 어느정도까지 허용해줄지 약간 의견이 다를수 있지만 말씀주신 걸로 이해해도 충분해 보입니다.
나중에 스프링을 배울 때 배우실텐데(그것도 처음엔 안배울겁니다.) Client서 보내온 데이터를 DTO(혹은 도메인)로 변경해주는건 Controller 앞단에서 스프링이 매핑해준답니다.
여러가지를 고려해봤을 때 순수하게 잘못된 입력에 의한 처리는 View에서 처리해도 충분하다고 판단되어 저는 View에서 이에 대한 처리를 하는걸 허용하는 편입니다.(선택은 크루들 몫이지만요 😋)

binghe819 replied:
꼼꼼한 피드백 감사합니다! :)

Hue9010 reviewed (approved):
마크! 이전 피드백 반영을 잘해주셨네요! 👍
이번 단계는 여기서 머지하도록 할게요.
다음 단계를 진행해주시면 될 것 같아요!
수고하셨습니다! 🙏
